packages/mongodb-ts-autocomplete/src/autocompleter.ts

```diff
@@ -0,0 +1,230 @@
+import autocompleteTypes from './fixtures/autocomplete-types';
+
+import type { JSONSchema } from 'mongodb-schema';
+import { toTypescriptTypeDefinition } from 'mongodb-schema';
+
+import {
+  inferCollectionNameFromFunctionCall,
+  compileSourceFile,
+} from './cdt-analyser';
+
+import { CachingAutocompletionContext } from './autocompletion-context';
+import type { AutocompletionContext } from './autocompletion-context';
+
+type MongoDBAutocompleterOptions = {
+  context: AutocompletionContext;
+};
+
+class DatabaseSchema {
+  private collectionSchemas: Record<string, JSONSchema | undefined>;
+
+  constructor() {
+    this.collectionSchemas = Object.create(null);
+  }
+
+  setCollectionNames(collectionNames: string[]): void {
+    // add the missing ones as undefined
+    for (const collectionName of collectionNames) {
+      if (!this.collectionSchemas[collectionName]) {
+        this.collectionSchemas[collectionName] = undefined;
+      }
+    }
+
+    // remove the ones that don't exist anymore
+    const knownCollectionNames = new Set(collectionNames);
+    for (const key of Object.keys(this.collectionSchemas)) {
+      if (!knownCollectionNames.has(key)) {
+        delete this.collectionSchemas[key];
+      }
+    }
+  }
+
+  setCollectionSchema(collectionName: string, schema: JSONSchema): void {
+    this.collectionSchemas[collectionName] = schema;
+  }
+
+  toTypescriptTypeDefinition(): string {
+    const collectionProperties = Object.entries(this.collectionSchemas).map(
+      ([collectionName, schema]) => {
+        const def: string = schema ? toTypescriptTypeDefinition(schema) : `{}`;
+        const lines = def.split(/\n/g).map((line) => `      ${line}`);
+        return `  '${collectionName}': {
+      schema: ${lines.join('\n').trim()}
+    };`;
+      },
+    );
+
+    return `{
+  ${collectionProperties.join('\n')}
+  }`;
+  }
+}
+
+class ConnectionSchema {
+  private readonly databaseSchemas: Record<string, DatabaseSchema>;
+
+  constructor() {
+    this.databaseSchemas = Object.create(null);
+  }
+
+  addDatabase(databaseName: string) {
+    if (!this.databaseSchemas[databaseName]) {
+      this.databaseSchemas[databaseName] = new DatabaseSchema();
+    }
+  }
+
+  setDatabaseCollectionNames(databaseName: string, collectionNames: string[]) {
+    this.addDatabase(databaseName);
+    this.databaseSchemas[databaseName].setCollectionNames(collectionNames);
+  }
+
+  addCollectionSchema(
+    databaseName: string,
+    collectionName: string,
+    collectionSchema: JSONSchema,
+  ) {
+    this.addDatabase(databaseName);
+    this.databaseSchemas[databaseName].setCollectionSchema(
+      collectionName,
+      collectionSchema,
+    );
+  }
+
+  toTypescriptTypeDefinition(): string {
+    const databaseProperties = Object.entries(this.databaseSchemas).map(
+      ([databaseName, schema]) => {
+        const def = schema.toTypescriptTypeDefinition();
+        return `'${databaseName}': ${def}`;
```

Comment:
ditto?

```suggestion
        return `${JSON.stringify(databaseName)}: ${def}`;
```

I mean, very generally you always want to `JSON.stringify()` when inserting JS values into JS source code, and then if you need to properly handle `__proto__` you might even want to do `JSON.parse(${JSON.stringify(JSON.stringify())})`